### PR TITLE
Polkadot Protocol > Protocol Components > Accounts > Account Data Structures

### PR DESCRIPTION
**Pull request by kapetan3sid**
Migrated https://docs.substrate.io/reference/account-data-structures/ page

### CODE REVIEW

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
+---
+title: Account Data Structure
+description: TODO
+---
+
+# Account Data Structure
+
+!!! warning "Update Notice"
```
Comment (CrackTheCode016) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:8]: Not sure if we need to keep this in these docs?

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    Please read Substrate to [Polkadot SDK](https://docs.substrate.io/polkadot-sdk/) page first.
+
+Accounts are a foundational element of any blockchain, and in Substrate, they serve various functions. This article explains how accounts are stored in Substrate and how their data structures are utilized to manage the account lifecycle within the runtime logic.
```
Comment (CrackTheCode016) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:11]: Someone else can correct me, but I think we are using `Polkadot SDK` and not `Substrate`.

```suggestion
Accounts are a foundational element of any blockchain. This page explains how accounts are managed in the Polkadot SDK and how their data structures are utilized to manage the account lifecycle within the runtime.
```

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+## Account
+
+The `Account` data type is a generically defined storage map within the [`frame-system`](https://paritytech.github.io/polkadot-sdk/master/src/frame_system/lib.rs.html) pallet:
+
+```rs
```
Comment (CrackTheCode016) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:17]: Not sure it is productive to have this snippet here, in the comment above I linked the struct itself in the codebase, which is more maintainable IMO.

Instead, a simple explanation of how it is a mapping of `AccountId` -> `AccountInfo` would suffice

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+- `T::AccountId` is used as the key for over the `AccountInfo<T::Nonce, T::AccountData>` struct.
+
+!!! note
+    See [StorageMap API](https://paritytech.github.io/polkadot-sdk/master/frame_support/storage/types/struct.StorageMap.html) for more details.
+
+## Account Info
+
+The `AccountInfo` for an account is also defined within the [`frame-system`](https://paritytech.github.io/polkadot-sdk/master/src/frame_system/lib.rs.html) pallet:
+
+```rs
```
Comment (CrackTheCode016) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:43]: If we do choose to leave code snippets, then we should put them in snippets, but I am more leaning towards just linking the line in the code base.

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+# Account Data Structure
+
+!!! warning "Update Notice"
+    Please read Substrate to [Polkadot SDK](https://docs.substrate.io/polkadot-sdk/) page first.
+
+Accounts are a foundational element of any blockchain, and in Substrate, they serve various functions. This article explains how accounts are stored in Substrate and how their data structures are utilized to manage the account lifecycle within the runtime logic.
+
+## Account
+
+The `Account` data type is a generically defined storage map within the [`frame-system`](https://paritytech.github.io/polkadot-sdk/master/src/frame_system/lib.rs.html) pallet:
```
Comment (CrackTheCode016) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:15]: IRT we link to the Polkadot SDK, or Rust Docs code?

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+  /// The number of modules that allow this account to exist for their own purposes only. The
+  /// account may not be reaped until this and `providers` are both zero.
+  pub sufficients: RefCount,
+  /// The additional data that belongs to this account. Used to store the balance(s) in a lot of
+  /// chains.
+  pub data: AccountData,
+}
+```
+
+Each account has an `AccountInfo` structure that includes the following components:
```
Comment (CrackTheCode016) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:63]: I think a list like this is probably better than including the struct, IMHO.

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+### Providers Reference Counters
+
+The `providers` reference counter indicates whether an account is ready to be depended upon. For example, it is incremented when a new account is created with a balance greater than the existential deposit.
+
+The `providers` reference counter prevents Substrate pallets from storing data about an account until the account is active (`providers` > 0)
+
+### Consumers Reference Counters
+
+The `consumers` reference counter ensures that Substrate pallets don't remove an account until all associated data across pallets is deleted (that is, when `consumers` == 0).
```
Comment (CrackTheCode016) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:87]:
```suggestion
The `consumers` reference counter ensures that other runtime pallets don't remove an account until all associated data across pallets is deleted (that is, when `consumers` == 0). 
```

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+The providers counter must be greater than zero for the consumers counter to be incremented.
+
+### Providers Reference Counters
+
+The `providers` reference counter indicates whether an account is ready to be depended upon. For example, it is incremented when a new account is created with a balance greater than the existential deposit.
+
+The `providers` reference counter prevents Substrate pallets from storing data about an account until the account is active (`providers` > 0)
+
+### Consumers Reference Counters
```
Comment (CrackTheCode016) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:85]: Not sure if necessary here, but maybe we have a bullet list explaining what a provider, consumer, and sufficient is in this context?

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+### Account Deactivation 
+
+Pallets also include cleanup functions that decrement the `providers` reference counter, marking the account as deactivated within the pallet-managed scope. 
+
+Once both the `providers` and `consumers` reference counters reach zero, the account is considered deactivated by all on-chain pallets.
+
+### Updating Counters
+
+Runtime developers can manage these counters using the `inc_consumers()`, `dec_consumers()`, `inc_providers()`, `dec_providers()`, `inc_sufficients()`, and `dec_sufficients()` methods provided by the `frame-system` pallet.
```
Comment (CrackTheCode016) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:105]: I am not sure if we should link the functions themselves, instead link where the functions live, i,e.,: https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method

Comment (CrackTheCode016) [reply polkadot-protocol/protocol-components/accounts/account-data-structure.md:105]: Although in this case, it is highly unlikely something like `frame_system` would change, at least a bullet list format would be better.

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+- `can_inc_consumer()` to check if an account is ready to be used (`providers` > 0).
+- `can_dec_provider()` to check if an account is no longer referenced in runtime whatsoever (`consumers` == 0) before decrementing providers to 0.
+- `is_provider_required()` to check if an account has outstanding consumer references (`consumers` > 0).
+
+!!! note
+    See [frame-system API](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html) for more details.
+
+## AccountData Trait and Implementation
+
+The `AccountInfo` can be any structure, provided it meets the requirements of the associated type `AccountData` trait bound defined in the [`frame-system::pallet::Config`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/trait.Config.html) trait.
```
Comment (CrackTheCode016) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:120]: Link AccountInfo via Polkadot SDK Docs

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+### Updating Counters
+
+Runtime developers can manage these counters using the `inc_consumers()`, `dec_consumers()`, `inc_providers()`, `dec_providers()`, `inc_sufficients()`, and `dec_sufficients()` methods provided by the `frame-system` pallet. 
+
+For every increment of a counter during an account's lifecycle, a corresponding decrement should be called to ensure proper management of the account state.
+
+There are also three query functions to ease usage on these counters:
+
+- `can_inc_consumer()` to check if an account is ready to be used (`providers` > 0).
```
Comment (CrackTheCode016) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:111]: ditto, or link the functions in the docs

**Review by CrackTheCode016 — CHANGES_REQUESTED** [review]
Overall looking good, some minor comments / things to be addressed

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md (hunk 3 above)
Comment (0xLucca) [reply polkadot-protocol/protocol-components/accounts/account-data-structure.md:17]: I agree

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md (hunk 1 above)
Comment (0xLucca) [reply polkadot-protocol/protocol-components/accounts/account-data-structure.md:8]: I think we should remove this

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md (hunk 9 above)
Comment (kapetan3sid) [reply polkadot-protocol/protocol-components/accounts/account-data-structure.md:105]: I created a bullet list and added descriptions to the functions listed

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md (hunk 10 above)
Comment (kapetan3sid) [reply polkadot-protocol/protocol-components/accounts/account-data-structure.md:120]: I linked AccountInfo to frame_system::AccountInfo , tell me if this is the correct linkl

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md (hunk 11 above)
Comment (kapetan3sid) [reply polkadot-protocol/protocol-components/accounts/account-data-structure.md:111]: Same, I added links to all these three functions from **frame_system::pallet::Pallet**

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md (hunk 8 above)
Comment (kapetan3sid) [reply polkadot-protocol/protocol-components/accounts/account-data-structure.md:85]: Not sure I understand what to add here, could you give me more context?

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+!!! note
+    See [frame-system API](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html) for more details.
+
+## AccountData Trait and Implementation
+
+The `AccountInfo` can be any structure, provided it meets the requirements of the associated type `AccountData` trait bound defined in the [`frame-system::pallet::Config`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/trait.Config.html) trait. 
+
+By default, the Substrate runtime configures `AccountInfo` as defined in the [`pallet-balances`](https://paritytech.github.io/polkadot-sdk/master/pallet_balances/struct.AccountData.html).
+
+## Where to Go Next
```
[the remark this answers is not in the document]
Comment (kapetan3sid) [reply polkadot-protocol/protocol-components/accounts/account-data-structure.md:102]: makes sense, will remove links

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 13]
```diff
@@ -0,0 +1,100 @@
+---
+title: Account Data Structure
+description: TODO
```
Comment (CrackTheCode016) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:3]: Description is missing

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md (hunk 8 above)
Comment (CrackTheCode016) [reply polkadot-protocol/protocol-components/accounts/account-data-structure.md:53]: Ah nvm, seems to be a list above that I missed, all good now!

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+- [`is_provider_required()`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.is_provider_required){target=\_blank} to check if an account has outstanding consumer references (`consumers` > 0).
+
+!!! note
+    See [frame-system API](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html){target=\_blank} for more details.
+
+## AccountData Trait and Implementation
+
+The [`AccountInfo`](https://paritytech.github.io/polkadot-sdk/master/frame_system/struct.AccountInfo.html){target=\_blank} can be any structure, provided it meets the requirements of the associated type `AccountData` trait bound defined in the [`frame-system::pallet::Config`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/trait.Config.html){target=\_blank} trait. 
+
+By default, the Substrate runtime configures `AccountInfo` as defined in the [`pallet-balances`](https://paritytech.github.io/polkadot-sdk/master/pallet_balances/struct.AccountData.html){target=\_blank}.
```
Comment (CrackTheCode016) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:100]: I've started to just say "runtime" instead "Substrate runtime" for consistency, @eshaben is this fine?

```suggestion
By default, the runtime configures `AccountInfo` as defined in the [`pallet-balances`](https://paritytech.github.io/polkadot-sdk/master/pallet_balances/struct.AccountData.html){target=\_blank}.
```

**Review by CrackTheCode016 — APPROVED** [review]
Besides some formatting stuff, LGTM content wise!

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+The `Account` data type is a generically defined storage map within the [`frame-system`](https://paritytech.github.io/polkadot-sdk/master/src/frame_system/lib.rs.html#900){target=\_blank} pallet:
+
+ --8<-- 'code/polkadot-protocol/protocol-components/accounts/account-id-information.md'
+
+The StorageMap for an Account consists of the following parameters:
+
+- The first parameter `(_)` is used in macro expansion.
+- `Blake2_128Concat` specifies the hashing algorithm to use.
+- `T::AccountId` is used as the key for over the `AccountInfo<T::Nonce, T::AccountData>` struct.
```
Comment (0xLucca) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:20]:
```suggestion
- The first parameter `(_)` is used in macro expansion
- `Blake2_128Concat` specifies the hashing algorithm to use
- `T::AccountId` is used as the key for over the `AccountInfo<T::Nonce, T::AccountData>` struct
```

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+--8<-- 'code/polkadot-protocol/protocol-components/accounts/account-info.md'
+
+Each account has an `AccountInfo` structure that includes the following components:
+
+- `nonce`: Tracks the number of transactions the account has sent.
+- `consumers`: A reference counter indicating the number of other modules that depend on this account's existence.
+- `providers`: A reference counter indicating the number of modules that permit this account to exist.
+- `sufficients`: A reference counter indicating the number of modules that allow this account to exist for their own purposes only.
+- `AccountData`: A configurable structure that can store various types of data specific to the account.
```
Comment (0xLucca) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:37]:
```suggestion
- `nonce` - tracks the number of transactions the account has sent
- `consumers` - a reference counter indicating the number of other modules that depend on this account's existence
- `providers` - a reference counter indicating the number of modules that permit this account to exist
- `sufficients` - a reference counter indicating the number of modules that allow this account to exist for their own purposes only
- `AccountData` - a configurable structure that can store various types of data specific to the account
```

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+### Updating Counters
+
+Runtime developers can manage these counters using the:
+
+-  `inc_consumers()` - Increment the reference counter on an account.
+- `dec_consumers()` - Decrement the reference counter on an account.
+- `inc_providers()` - Increment the provider reference counter on an account.
+- `dec_providers()` - Decrement the provider reference counter on an account.
+- `inc_sufficients()` - Increment the self-sufficient reference counter on an account.
+- `dec_sufficients()` - Decrement the sufficients reference counter on an account.
```
Comment (0xLucca) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:80]:
```suggestion
-  `inc_consumers()` - increment the reference counter on an account
- `dec_consumers()` - decrement the reference counter on an account
- `inc_providers()` - increment the provider reference counter on an account
- `dec_providers()` - decrement the provider reference counter on an account
- `inc_sufficients()` - increment the self-sufficient reference counter on an account
- `dec_sufficients()` - decrement the sufficients reference counter on an account
```

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+!!!note
+    All these methods are provided by the [`frame-system`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method){target=\_blank} pallet.
+
+For every increment of a counter during an account's lifecycle, a corresponding decrement should be called to ensure proper management of the account state.
+
+There are also three query functions to ease usage on these counters:
+
+- [`can_inc_consumer()`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.can_inc_consumer){target=\_blank} to check if an account is ready to be used (`providers` > 0).
+- [`can_dec_provider()`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.can_dec_provider){target=\_blank} to check if an account is no longer referenced in runtime whatsoever (`consumers` == 0) before decrementing providers to 0.
+- [`is_provider_required()`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.is_provider_required){target=\_blank} to check if an account has outstanding consumer references (`consumers` > 0).
```
Comment (0xLucca) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:91]:
```suggestion
- [`can_inc_consumer()`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.can_inc_consumer){target=\_blank} to check if an account is ready to be used (`providers` > 0)
- [`can_dec_provider()`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.can_dec_provider){target=\_blank} to check if an account is no longer referenced in runtime whatsoever (`consumers` == 0) before decrementing providers to 0
- [`is_provider_required()`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.is_provider_required){target=\_blank} to check if an account has outstanding consumer references (`consumers` > 0)
```

**Review by 0xLucca — COMMENTED** [review]
Fix this minor stuff and Bader's changes, and should be fine to merge

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+
+Pallets also include cleanup functions that decrement the `providers` reference counter, marking the account as deactivated within the pallet-managed scope. 
+
+Once both the `providers` and `consumers` reference counters reach zero, the account is considered deactivated by all on-chain pallets.
+
+### Updating Counters
+
+Runtime developers can manage these counters using the:
+
+-  `inc_consumers()` - Increment the reference counter on an account.
```
Comment (CrackTheCode016) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:75]:
```suggestion
- `inc_consumers()` - Increment the reference counter on an account.
```

**Review by CrackTheCode016 — APPROVED** [review]
lgtm, except for one super tiny nit

**Review by 0xLucca — DISMISSED** [review]
LGTM

**Review by CrackTheCode016 — APPROVED** [review]
For now this is fine, just need to make sure we come back to add some extra context later.

**Comment by dawnkelly09** [comment]
@eshaben I did make a few edits so please do a final pass over this.

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 20]
```diff
@@ -28,7 +28,7 @@ The `StorageMap` consists of the following parameters:
 - **`ValueQuery`** - a trait that defines how queries to the storage map behave when no value is found, returning the default value if a query for a key returns no result (instead of `None`)
 
 !!! note
-    See the [`StorageMap` API](https://paritytech.github.io/polkadot-sdk/master/frame_support/storage/types/struct.StorageMap.html){target=\_blank} for more details.
+    See the [`StorageMap` rustdocs](https://paritytech.github.io/polkadot-sdk/master/frame_support/storage/types/struct.StorageMap.html){target=\_blank} for more details.
```
Comment (eshaben) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:31]: I can't leave a suggestion but I think should be Rust docs

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 21]
```diff
@@ -101,7 +101,7 @@ There are also three query functions to ease usage on these counters:
 - [**`is_provider_required()`**](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.is_provider_required){target=\_blank} to check if an account has outstanding consumer references (`consumers` greater than 0)
 
 !!! note
-    See the [System pallet API](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html){target=\_blank} for more details.
+    See the [System pallet rustdocs](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html){target=\_blank} for more details.
```
Comment (eshaben) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:104]: Same here - Rust docs

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+- **`Blake2_128Concat`** - the hashing function used for keys in the storage map
+- **`T::AccountId`** - the key type for the storage map, representing the account ID, which is used to access the corresponding `AccountInfo<T::Nonce, T::AccountData>` struct in storage
+- `AccountInfo<T::Nonce, T::AccountData>` - the value type stored in the map. For each account ID, the map stores an `AccountInfo` struct containing:
+    - **`T::Nonce`** - a nonce for the account, which is incremented with each transaction to ensure transaction uniqueness
+    - **`T::AccountData`** - custom account data defined by the runtime configuration, which could include balances, locked funds, or other relevant information
+- **`ValueQuery`** - a trait that defines how queries to the storage map behave when no value is found, returning the default value if a query for a key returns no result (instead of `None`)
+
+!!! note
+    See the [`StorageMap` rustdocs](https://paritytech.github.io/polkadot-sdk/master/frame_support/storage/types/struct.StorageMap.html){target=\_blank} for more details.
```
Comment (eshaben) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:31]:
```suggestion
    See the [`StorageMap` Rust docs](https://paritytech.github.io/polkadot-sdk/master/frame_support/storage/types/struct.StorageMap.html){target=\_blank} for more details.
```

File: polkadot-protocol/protocol-components/accounts/account-data-structure.md [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+For every counter increment during an account's lifecycle, a corresponding decrement should be called to ensure proper management of the account state.
+
+There are also three query functions to ease usage on these counters:
+
+- [**`can_inc_consumer()`**](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.can_inc_consumer){target=\_blank} - to check if an account is ready to be used (`providers` greater than 0)
+- [**`can_dec_provider()`**](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.can_dec_provider){target=\_blank} - to check if an account is no longer referenced in the runtime whatsoever (`consumers` equal 0) before decrementing `providers` to 0
+- [**`is_provider_required()`**](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.is_provider_required){target=\_blank} to check if an account has outstanding consumer references (`consumers` greater than 0)
+
+!!! note
+    See the [System pallet rustdocs](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html){target=\_blank} for more details.
```
Comment (eshaben) [inline polkadot-protocol/protocol-components/accounts/account-data-structure.md:104]:
```suggestion
    See the [System pallet Rust docs](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html){target=\_blank} for more details.
```

**Review by eshaben — APPROVED** [review]
LGTM

**Review by nhussein11 — APPROVED** [review]
LGTM!

**Pull request merged**